src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+
+namespace BlazorWasm_CSharp.Shared.Models
+{
+    public class UserInfo
+    {
+        public static readonly UserInfo Anonymous = new UserInfo();
+
+        public UserInfo() { }
+
+        public UserInfo(ClaimsPrincipal claimsPrincipal)
+        {
+            var identity = claimsPrincipal.Identity;
+            AuthenticationType = identity.AuthenticationType;
+            if (identity is ClaimsIdentity claimsIdentity)
+            {
+                NameClaimType = claimsIdentity.NameClaimType;
+                RoleClaimType = claimsIdentity.RoleClaimType;
+            }
+            else
+            {
+                NameClaimType = "name";
+                RoleClaimType = "roles";
+            }
+
+            if (claimsPrincipal.Claims.Any())
+            {
+                var claims = new List<ClaimValue>();
+                foreach (var claim in claimsPrincipal.Claims)
+                {
+                    claims.Add(new ClaimValue(claim.Type, claim.Value));
+                }
+
+                Claims = claims;
+            }
+        }
+
+        public string AuthenticationType { get; set; }
+
+        public string NameClaimType { get; set; }
+
+        public string RoleClaimType { get; set; }
+
+        public ICollection<ClaimValue> Claims { get; set; }
+    }
```

Comment:
I'd be interested in restructuring this class a bit to make it clearer that we *don't* expect the server to disclose everything it knows about the `ClaimsPrincipal`. That is,

 * By default we should only expose the "name" claim
 * By default we do not expose `AuthenticationType`, since the client has no use for this info by default. Developers should manually add some code (e.g., the whole `AuthenticationType` property) if they want to communicate this to the client
 * By default we shouldn't expose any other claims, as they might contain sensitive info. We could have a comment telling developers that if they want to, they can write code to populate more of the `Claims` collection using data from `claimsPrincipal`, but it's up to them to decide carefully which claim data is safe to expose to end-users. For example, don't expose a claim if it contains a token that grants access to some other service which is only supposed to be called from the server.

So by default, `UserInfo` might be reducible just to having the `Claims` property, in which by default we only place the "name" claim.

### About roles

I'm slightly less sure about this, but I suspect that similarly we shouldn't assume they want to expose the full list to the client, or even that client-side roles have the same names as server-side ones. We might have a commented-out block of code giving an example of how you can specify a role if you want, e.g.,

```cs
// To pass information about roles to the client-side application, add boolean properties for each client-side
// role and use logic similar to the following to populate them. Note that typically you do not need to expose
// the full list of roles to the client.
// IsSuperAdmin = claimsPrincipal.IsInRole("IsSuperAdmin");
```

Reply:
> * By default we shouldn't expose any other claims, as they might contain sensitive info.
The default set of claims doesn't include anything sensitive.

> We could have a comment telling developers that if they want to, they can write code to populate more of the `Claims` collection using data from `claimsPrincipal`, but it's up to them to decide carefully which claim data is safe to expose to end-users. 

I think we could have a comment in the other direction saying that if you want to, you can customize the list of claims to share with the user here. Even if you add properties to the IdentityUser they don't get exposed as claims, so I think it is fine.

> For example, don't expose a claim if it contains a token that grants access to some other service which is only supposed to be called from the server.

I don't think this is a valid a valid example. We don't store access tokens inside user claims and our guidance would be not to store tokens in a claim inside the identity. Basically, treat secrets as secrets and don't mix them with any other information.

Also, in this scenario you control the app E2E, so you are only sharing this information with your app.

> ### About roles
> I'm slightly less sure about this, but I suspect that similarly we shouldn't assume they want to expose the full list to the client, or even that client-side roles have the same names as server-side ones. We might have a commented-out block of code giving an example of how you can specify a role if you want, e.g.,

Roles are not secrets.

I think exposing roles enables a better experience by default, as you are able to use them in your pages to limit access directly through the UI to some areas of your app with the appropriate Authorize attribute.

My reasoning behind all these decisions are:
 * You control your app, end to end
 * The authentication implementation is meant to be used with Identity.
 * There is already a place to opt-in to adding claims which defined by Identity (IUserClaimsPrincipalFactory)
    * If you add a claim to your app, normally you want to allow your app to consume it.

Reply:
This might all be fine, but would it be OK if we raise this in particular when we do security review? My concern is that we're creating a new form of hazard, where in the past the contents of a ClaimsIdentity would not be exposed to the user in plaintext (right?), but now it is. Even if we don't expect someone to put in things that should be secret from the user themselves, them might do so and not realise what's happening here. Examples would be:

 * A role the user should not be allowed to discover they are in (e.g., `HellbannedUsers`)
 * A claim that contains tokens that are only intended for use by the server, not client, even if we don't recommend this

Perhaps these are not valid scenarios but is it OK if we lay this out directly in security review for feedback there? I'm not saying this PR should be blocked on that, since it's only a very small part of this PR and we can tweak it later if needed.

> The authentication implementation is meant to be used with Identity.

I thought our goal was to make it agnostic to the server-side authentication mechanism. There's no reason Blazor WebAssembly should prefer people to use Identity in particular, or any other authentication mechanism. Anything should work as long as it tracks state in cookies.

Reply:
> * A role the user should not be allowed to discover they are in (e.g., `HellbannedUsers`)

I don't think roles are meant to be used that way. Normally they include you in a group, not exclude you from. As in, you have Admin role, you can access this area of the application and so on. 

What is described in this sample doesn't work with the way authorization works in ASP.NET Core either, as the Authorize attribute allows you to require a role, but not to avoid having it.

> I thought our goal was to make it agnostic to the server-side authentication mechanism. There's no reason Blazor WebAssembly should prefer people to use Identity in particular, or any other authentication mechanism. Anything should work as long as it tracks state in cookies.

I think we should have a very clear scope here. The template uses identity and we should not try to build anything more generic than that, based on the assumption that the templates are a starting point and that if you want to replace Identity with your own solution then you have all the power to modify the code in whatever way you see fit.

Reply:
This is a bit tangential, but hellbanning doesn't mean exclusion from anything. It's a flag associated with a user meaning that their actions will be treated as silent from the point of view of other users. For example, if you were hellbanned on GitHub, then you could still do all the normal stuff, but when you post comments or PRs, the system would see you were in the "hellbanned" group and flag your post/PR so it's invisible to other users. There are other ways to model this state, but someone *might* reasonably use a role for it without understanding how we're now disclosing that info.

Anyway let's delay the discussion about whether to list all roles until we do security review. We don't need to block this PR on it.